### PR TITLE
[WIP] make unicode vs bytes more distinct in py2

### PR DESCRIPTION
**Pull request by remdragon**
Implementing and testing the changes suggested by gvanrossum where py2's Text is redefined as Union[bytes,unicode] and automatic str->unicode promotion is disabled.

This PR is a WIP but submitting now for CI and peer review

See the following issues for more discussion:

https://github.com/python/typing/issues/208
https://github.com/python/typeshed/issues/2609

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
FYI there's a merge conflict.

File: stdlib/2/builtins.pyi [hunk 1]
```diff
@@ -66,18 +66,18 @@ class classmethod(object):  # Special, only valid as a decorator.
 
 class type(object):
     __bases__ = ...  # type: Tuple[type, ...]
-    __name__ = ...  # type: str
-    __module__ = ...  # type: str
+    __name__ = ...  # type: Text
```
Comment (JelleZijlstra) [inline stdlib/2/builtins.pyi]: I don't think this is correct. `__name__` is always str in Python 2, not unicode.

**Comment by JelleZijlstra** [comment]
And now there's another merge conflict and a CI failure. I'm also not sure most of the changes I checked are correct; sometimes it's really just `str`.

**Comment by remdragon** [comment]
I tried a couple ways and can't figure out how to overwrite type's __name__ attribute with a unicode string, so I believe you're right, I'll make that adjustment and fix the merge conflict.

**Comment by srittau** [comment]
I am going to close this for now. Please reopen if you should find to work on this.

**Pull request closed without merging**